bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/PageChangeListener.java

```diff
@@ -0,0 +1,164 @@
+     * @param itemUIRegistry the ItemUIRegistry which is needed for the functionality
+     * @param widgets the list of widgets that are part of the page.
+     */
+    public PageChangeListener(String sitemapName, String pageId, ItemUIRegistry itemUIRegistry, EList<Widget> widgets) {
+        this.sitemapName = sitemapName;
+        this.pageId = pageId;
+        this.itemUIRegistry = itemUIRegistry;
+        this.widgets = widgets;
+        items = getAllItems(widgets);
+        for (Item item : items) {
+            if (item instanceof GenericItem) {
+                ((GenericItem) item).addStateChangeListener(this);
+            } else if (item instanceof GroupItem) {
+                ((GroupItem) item).addStateChangeListener(this);
+            }
+        }
+    }
+
+    public String getSitemapName() {
+        return sitemapName;
+    }
+
+    public String getPageId() {
+        return pageId;
+    }
+
+    public void addCallback(SitemapSubscriptionCallback callback) {
+        callbacks.add(callback);
+    }
+
+    public void removeCallback(SitemapSubscriptionCallback callback) {
+        callbacks.remove(callback);
+    }
+
+    /**
+     * Disposes this instance and releases all resources.
+     */
+    public void dispose() {
+        for (Item item : items) {
+            if (item instanceof GenericItem) {
+                ((GenericItem) item).removeStateChangeListener(this);
+            } else if (item instanceof GroupItem) {
+                ((GroupItem) item).removeStateChangeListener(this);
+            }
+        }
+    }
+
+    /**
+     * Collects all items that are represented by a given list of widgets
+     *
+     * @param widgets
+     *            the widget list to get the items for added to all bundles containing REST resources
+     * @return all items that are represented by the list of widgets
+     */
+    private Set<Item> getAllItems(EList<Widget> widgets) {
+        Set<Item> items = new HashSet<Item>();
+        if (itemUIRegistry != null) {
+            for (Widget widget : widgets) {
+                String itemName = widget.getItem();
+                if (itemName != null) {
+                    try {
+                        Item item = itemUIRegistry.getItem(itemName);
+                        items.add(item);
+                    } catch (ItemNotFoundException e) {
+                        // ignore
+                    }
+                } else {
+                    if (widget instanceof Frame) {
+                        items.addAll(getAllItems(((Frame) widget).getChildren()));
+                    }
+                }
+            }
+        }
+        return items;
+    }
+
+    @Override
+    public void stateChanged(Item item, State oldState, State newState) {
+        Set<SitemapEvent> events = constructSitemapEvents(item, oldState, newState, widgets);
+        for (SitemapEvent event : events) {
+            // we transform the list of callbacks to a set in order to remove duplicates
```

Comment:
If we don't care about the above comment, we should create the set only once before the for-each-event loop and not in every iteration.